File: backend/services/account-service/internal/dto/fortune_card.go
package dto

import "time"

// MaxPageSize 是福卡各列表接口 pageSize 的上限，取值与 platform/api.MaxPageSize 一致。
//
// 与 coupon-service 的 dto.MaxPageSize 同一个理由：前端「要全集」的统一参数
// （admin-web 的 FULL_PAGE_PARAMS，pageSize=200）在更小的上限上会直接 400，然后把
// 一次翻译静默变成一列 ID。HTTP 层交给 api.ParsePage 校验，服务层与仓储层的兜底
// 也用它，避免漂移成三个数字。
const MaxPageSize = 200

// EntryQuery 是福卡流水查询的过滤条件。零值全部表示「不限」。
//
// From/To 是闭开区间 [From, To)：同一条流水只会落在相邻两天的其中一边，导出与对账
// 拼接时不会漏也不会重。指针而不是零值时间，是为了区分「没传」与「传了零时刻」。
type EntryQuery struct {
	UserID    string
	OrderNo   string
	EntryType string
	From      *time.Time
	To        *time.Time
	Page      int
	PageSize  int
}

// EntryResponse 是流水对外的一行。
//
// 字段名与 admin-web / 小程序读的是同一套（json tag 就是契约）：明细页要显示的
// 「变动后余额」是 BalanceAfter，不是当前余额——冲正重放时，只有它能回答「当时是多少」。
type EntryResponse struct {
	ID              string    `json:"id"`
	UserID          string    `json:"userId"`
	EntryType       string    `json:"entryType"`
	Amount          int64     `json:"amount"`
	BalanceAfter    int64     `json:"balanceAfter"`
	Title           string    `json:"title"`
	ReferenceType   string    `json:"referenceType"`
	ReferenceID     string    `json:"referenceId"`
	ReferenceNo     string    `json:"referenceNo"`
	ReversesEntryID string    `json:"reversesEntryId"`
	Remark          string    `json:"remark"`
	OccurredAt      time.Time `json:"occurredAt"`
	CreatedAt       time.Time `json:"createdAt"`
}

// AccountResponse 是一个用户的福卡账户。
//
// 没有账户行时余额是 0 而不是 404：账户行是第一次发放时才懒创建的，一个从没收到过
// 福卡的用户查自己的余额，答案是「0 张」这件事本身，不是「查不到这个人」。
//
// 三个余额一起给：客服最常问的是「我明明有 2 张，为什么抽奖说不够」——答案在
// FrozenBalance 上（有一张退款申请正冻着），只看 Balance 会看不出这句话哪里错了。
// AvailableBalance 是 Balance - FrozenBalance，算好了给，免得每个消费方各自减一次。
type AccountResponse struct {
	UserID           string    `json:"userId"`
	Balance          int64     `json:"balance"`
	FrozenBalance    int64     `json:"frozenBalance"`
	AvailableBalance int64     `json:"availableBalance"`
	CreatedAt        time.Time `json:"createdAt"`
	UpdatedAt        time.Time `json:"updatedAt"`
	// HasAccount 区分「有过账户、余额为 0」与「从来没有过账户」。客服看到
	// 「0 张」时会想知道是哪一种；两者在流水页上差别很大（一个有历史，一个没有）。
	HasAccount bool `json:"hasAccount"`
}

// FreezeQuery 是冻结核对的过滤条件。零值全部表示「不限」。
type FreezeQuery struct {
	UserID   string
	OrderNo  string
	Status   string
	Page     int
	PageSize int
}

// FreezeResponse 是冻结对外的一行。
//
// EntryKeys 一起给出来：客服判断「冻的是哪张卡」要看它（base 是订单基础赠送、
// bonus:{campaignId} 是加购加赠），而按退款范围冻的语义全在这几个键里。
type FreezeResponse struct {
	ID          string     `json:"id"`
	UserID      string     `json:"userId"`
	AfterSaleNo string     `json:"afterSaleNo"`
	OrderID     string     `json:"orderId"`
	OrderNo     string     `json:"orderNo"`
	EntryKeys   []string   `json:"entryKeys"`
	Amount      int64      `json:"amount"`
	Status      string     `json:"status"`
	Reason      string     `json:"reason"`
	OccurredAt  time.Time  `json:"occurredAt"`
	ReleasedAt  *time.Time `json:"releasedAt"`
	CreatedAt   time.Time  `json:"createdAt"`
	UpdatedAt   time.Time  `json:"updatedAt"`
}

// MiniappCardsResponse 是小程序端一次拿全的回答：余额 + 一页流水。
//
// 合成一个响应是因为明细页永远两个都要——分成两个接口只会让那一次渲染发两个请求，
// 还得处理「余额是新的、流水是旧的」这种半新半旧的状态。
type MiniappCardsResponse struct {
	Balance int64 `json:"balance"`
	// FrozenBalance / AvailableBalance 见 AccountResponse 的注释。小程序端页面上要写的是
	// 「可用 1 张（1 张因退款申请冻结中）」，而不是一个对不上账的余额。
	FrozenBalance    int64            `json:"frozenBalance"`
	AvailableBalance int64            `json:"availableBalance"`
	Entries          []*EntryResponse `json:"entries"`
	Total            int              `json:"total"`
	Page             int              `json:"page"`
	PageSize         int              `json:"pageSize"`
}

File: admin-web/src/services/fortuneCard.ts
import { request } from '@umijs/max';
import type { PageQuery, PageResult } from './pagination';

/**
 * 福卡账户域的接口（account-service）。字段名与后端 dto（internal/dto/fortune_card.go）
 * 的 json tag 逐字对应，同时也是页面上的 dataIndex——改一处要同批改三处。
 *
 * 张数一律是整数，没有小数也没有单位换算：福卡是抽奖凭证，不是金额。别把它塞进
 * money.ts 的那些格式化函数里。
 */

// ——— 枚举 ———

/**
 * 流水的收支类型。取值来自 account-service 的 model 常量，与
 * migrations/account/001 的 CHECK 约束一致。
 *
 * 加新码而这里没登记时，界面退回显示原始码（同 order.ts 的规矩）。
 */
export type FortuneCardEntryType = 'grant' | 'draw' | 'reverse';

/** 这条流水指向什么：订单赠送 / 参与抽奖 / 冲正某一条流水。 */
export type FortuneCardReferenceType = 'order' | 'draw' | 'entry' | '';

// ——— 形状 ———

/**
 * 一条福卡流水。
 *
 * `amount` 是**有符号**的：发放为正，抽奖扣减与冲正为负。页面因此不能按「数量」渲染，
 * 要带着符号显示，否则一次抽奖看起来像又发了一张。
 */
export type FortuneCardEntry = {
  id: string;
  userId: string;
  entryType: FortuneCardEntryType;
  amount: number;
  /**
   * 这笔发生**之后**的余额，不是当前余额。明细页要显示的就是它——冲正重放时只有它能
   * 回答「当时是多少」。要当前余额请看 getFortuneCardAccount。
   */
  balanceAfter: number;
  /** 小程序明细上那一行文案（后端生成，界面原样显示）。 */
  title: string;
  referenceType: FortuneCardReferenceType;
  referenceId: string;
  /** 人可读的号：订单赠送时是订单号。按它筛流水就是靠这个字段。 */
  referenceNo: string;
  /** 冲正时指向被冲的那一条；其余为空串。 */
  reversesEntryId: string;
  remark: string;
  /** 业务发生时间（订单完成时间），与 created_at 可能不同。 */
  occurredAt: string;
  createdAt: string;
};

/** 冻结状态。取值来自 account-service 的 model 常量，与 migrations/account/003 的 CHECK 一致。 */
export type FortuneCardFreezeStatus = 'frozen' | 'released';

/**
 * 一条退款冻结：一张售后单冻住的那几笔发放。
 *
 * 冻结**不是账变**——balance 与流水一格不动，被冻的卡只是不能拿去抽奖。所以这里的
 * `amount` 不是「余额减少了几张」，别和 FortuneCardEntry.amount 混着读。解冻之后
 * `amount` 保留（审计要看当初冻了多少），`status` 变成 released。
 */
export type FortuneCardFreeze = {
  id: string;
  userId: string;
  /** 发起这次冻结的售后单号。它也是后端那把幂等键。 */
  afterSaleNo: string;
  orderId: string;
  orderNo: string;
  /**
   * 被冻的那几笔发放的幂等键（`order:{id}:base` / `order:{id}:bonus:{campaignId}`）。
   * 「按退款范围冻」的语义全在这里：只退加购行时这里只有 bonus 那一个键。
   */
  entryKeys: string[];
  /** 实际冻住的张数。可能是 0（申请早于发放，或卡已经被抽掉了）。 */
  amount: number;
  status: FortuneCardFreezeStatus;
  /** 后端生成的一句话：退款申请中 / 被驳回 / 用户撤销。 */
  reason: string;
  occurredAt: string;
  /** 只有 status=released 时有值。 */
  releasedAt: string | null;
  createdAt: string;
  updatedAt: string;
};

/** 一个用户的福卡账户。 */
export type FortuneCardAccount = {
  userId: string;
  /** 总余额（含被冻的）。明细页上「变动后余额」对的就是它。 */
  balance: number;
  /** 退款冻结中的张数。 */
  frozenBalance: number;
  /** 能拿去抽奖的张数 = balance - frozenBalance。后端算好了给，别在前端再减一次。 */
  availableBalance: number;
  createdAt: string;
  updatedAt: string;
  /**
   * 区分「有过账户、余额为 0」与「从来没有过账户」。两者都是 balance 0，但流水页上
   * 差别很大（一个有历史，一个一条都没有），所以文案要分开写。
   */
  hasAccount: boolean;
};

// ——— 查询参数 ———

/**
 * 流水筛选。后端 EntryQuery 里全是精确匹配，`orderNo` 也不例外（所以调用前 trim）。
 *
 * `from`/`to` 是**闭开区间** [from, to)：同一条流水只会落在相邻两天的其中一边，
 * 拼接时不会漏也不会重。必须传带时区的 RFC3339（用 toRFC3339 转，与订单列表同款）。
 */
export type FortuneCardEntryQuery = PageQuery & {
  userId?: string;
  orderNo?: string;
  entryType?: FortuneCardEntryType;
  from?: string;
  to?: string;
};

/**
 * 冻结筛选。三个都是精确匹配（后端 dto.FreezeQuery），所以调用前 trim；
 * `status` 不传就是「不限」，界面上的「全部」要翻译成不传，而不是传空串。
 */
export type FortuneCardFreezeQuery = PageQuery & {
  userId?: string;
  orderNo?: string;
  status?: FortuneCardFreezeStatus;
};

// ——— 接口 ———

/**
 * 按条件查福卡流水（后台）。订单详情的「福卡」tab 就是按 orderNo 用它。
 *
 * 后端把 `/v1/admin/fortune-cards/entries` 注册在 `/v1/admin/fortune-cards/{userId}`
 * **之前**，所以这里不会被打成「userId 是 entries」——两者在网关与路由上都分得开。
 */
export async function listFortuneCardEntries(params?: FortuneCardEntryQuery) {
  return request<PageResult<FortuneCardEntry>>('/api/v1/admin/fortune-cards/entries', {
    params,
  });
}

/**
 * 查某个用户的福卡账户（客服最常问的那一句：他还有几张）。
 *
 * 没有账户行时回的是余额 0 + hasAccount=false，**不是 404**——账户行是第一次发放时才
 * 懒创建的，一个从没收到过福卡的用户查余额，答案就是「0 张」。
 */
export async function getFortuneCardAccount(userId: string) {
  return request<FortuneCardAccount>(`/api/v1/admin/fortune-cards/${userId}`);
}

/**
 * 按条件查退款冻结（后台）。订单详情的「福卡」tab 用它显示「冻结中 N 张」。
 *
 * 与 entries 同一条路由规矩：后端把 `/v1/admin/fortune-cards/freezes` 也注册在
 * `/{userId}` **之前**。名字里的 Freezes 不能省——`/fortune-cards/{userId}` 那条会把
 * 路径段当 userId 解析，写成 `/fortune-cards/freezes` 会被打成「userId 不是 UUID」的 400。
 */
export async function listFortuneCardFreezes(params?: FortuneCardFreezeQuery) {
  return request<PageResult<FortuneCardFreeze>>('/api/v1/admin/fortune-cards/freezes', {
    params,
  });
}
